gax/src/main/java/com/google/api/gax/rpc/Watchdog.java

```diff
@@ -134,26 +143,37 @@ private void runUnsafe() {
   @Override
   public void shutdown() {
     future.cancel(false);
+    // Unregister the main thread
+    phaser.arriveAndDeregister();
```

Comment:
Might be good to avoid double calls:

if (future.cancel(true)) {
phaser.arriveAndDeregister();

}